File: backend/app/models/trip.py
from sqlmodel import SQLModel, Field, Relationship, Column
from typing import Optional, List
from datetime import datetime, date
from sqlalchemy.dialects.postgresql import JSONB


class Trip(SQLModel, table=True):
    """User trip itineraries"""
    __tablename__ = "trips"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: str = Field(foreign_key="user_profiles.id", index=True)
    title: str = Field(max_length=200)
    summary: Optional[str] = None
    destination: str = Field(index=True, max_length=100)
    days: int = Field(ge=1, le=30)
    budget: Optional[str] = Field(default="mid", max_length=20)
    travel_style: List[str] = Field(default_factory=list, sa_column=Column(JSONB, default=list))
    companions: Optional[str] = Field(default="solo", max_length=20)
    time_of_year: Optional[str] = Field(default=None, max_length=50)
    total_budget_min: Optional[int] = None
    total_budget_max: Optional[int] = None
    is_public: bool = Field(default=False)
    view_count: int = Field(default=0)
    created_from_template_id: Optional[int] = None
    created_at: datetime = Field(default_factory=datetime.utcnow, index=True)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class TripDay(SQLModel, table=True):
    """Individual days within a trip"""
    __tablename__ = "trip_days"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    trip_id: int = Field(foreign_key="trips.id", index=True)
    day_number: int = Field(ge=1)
    title: str = Field(max_length=200)
    trip_date: Optional[date] = None
    total_cost_min: Optional[int] = None
    total_cost_max: Optional[int] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


class Activity(SQLModel, table=True):
    """Activities within a trip day"""
    __tablename__ = "activities"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    trip_day_id: int = Field(foreign_key="trip_days.id", index=True)
    name: str = Field(max_length=200)
    description: Optional[str] = None
    category: str = Field(max_length=50)
    place_id: Optional[str] = Field(default=None, foreign_key="places.id", index=True)
    latitude: Optional[float] = Field(default=None, ge=-90, le=90)
    longitude: Optional[float] = Field(default=None, ge=-180, le=180)
    address: Optional[str] = None
    duration_minutes: Optional[int] = Field(default=None, ge=0)
    cost_min: Optional[int] = None
    cost_max: Optional[int] = None
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    booking_required: bool = Field(default=False)
    booking_url: Optional[str] = None
    notes: Optional[str] = None
    sort_order: int = Field(default=0)
    created_at: datetime = Field(default_factory=datetime.utcnow)


# Response models
class TripPublic(SQLModel):
    id: int
    title: str
    summary: Optional[str]
    destination: str
    days: int
    budget: Optional[str]
    created_at: datetime


class TripDetail(TripPublic):
    total_budget_min: Optional[int]
    total_budget_max: Optional[int]
    is_public: bool
    view_count: int


class TripDayDetail(SQLModel):
    day_number: int
    title: str
    trip_date: Optional[date]


class ActivityDetail(SQLModel):
    name: str
    description: Optional[str]
    category: str
    latitude: Optional[float]
    longitude: Optional[float]
    start_time: Optional[str]
    duration_minutes: Optional[int]

File: frontend/atla/src/types/trip.ts
/**
 * Trip Type Definitions
 * 
 * These types match the backend SQLModel/Pydantic models exactly.
 * Used for type-safe API communication and component props.
 */

/**
 * Activity within a trip day
 * Matches backend: app/models/trip.py - Activity
 */
export interface Activity {
  id: number
  trip_day_id: number
  name: string
  description?: string
  category: 'attraction' | 'restaurant' | 'accommodation' | 'transport'
  place_id?: string
  latitude?: number
  longitude?: number
  address?: string
  duration_minutes?: number
  cost_min?: number
  cost_max?: number
  start_time?: string  // Format: "HH:MM"
  end_time?: string
  booking_required: boolean
  booking_url?: string
  notes?: string
  sort_order: number
  created_at: string  // ISO date string
}

/**
 * Day within a trip
 * Matches backend: app/models/trip.py - TripDay
 */
export interface TripDay {
  id: number
  trip_id: number
  day_number: number
  title: string
  trip_date?: string  // ISO date string (YYYY-MM-DD)
  total_cost_min?: number
  total_cost_max?: number
  created_at: string
  activities: Activity[]
}

/**
 * Trip (itinerary)
 * Matches backend: app/models/trip.py - Trip
 */
export interface Trip {
  id: number
  user_id: string
  title: string
  summary?: string
  destination: string
  days: number
  budget?: 'low' | 'mid' | 'luxury'
  travel_style: string[]
  companions?: 'solo' | 'couple' | 'friends' | 'family'
  time_of_year?: string
  total_budget_min?: number
  total_budget_max?: number
  is_public: boolean
  view_count: number
  created_from_template_id?: number
  created_at: string
  updated_at: string
}

/**
 * Trip with nested days and activities
 * Returned by GET /api/trips/{id} with full details
 */
export interface TripWithDetails extends Trip {
  trip_days: TripDay[]
}

/**
 * Public trip view (minimal info for public listings)
 * Matches backend: TripPublic Pydantic model
 */
export interface TripPublic {
  id: number
  title: string
  summary?: string
  destination: string
  days: number
  budget?: string
  created_at: string
}

/**
 * Create trip request body
 * Matches backend: TripCreate Pydantic model
 */
export interface CreateTripRequest {
  title: string
  summary?: string
  destination: string
  days: number
  budget?: 'low' | 'mid' | 'luxury'
  travel_style?: string[]
  companions?: 'solo' | 'couple' | 'friends' | 'family'
  time_of_year?: string
  total_budget_min?: number
  total_budget_max?: number
  is_public?: boolean
}

/**
 * Update trip request body
 * Matches backend: TripUpdate Pydantic model
 */
export interface UpdateTripRequest {
  title?: string
  summary?: string
  destination?: string
  days?: number
  budget?: 'low' | 'mid' | 'luxury'
  travel_style?: string[]
  companions?: 'solo' | 'couple' | 'friends' | 'family'
  time_of_year?: string
  total_budget_min?: number
  total_budget_max?: number
  is_public?: boolean
}

/**
 * Activity detail (simplified for responses)
 * Matches backend: ActivityDetail Pydantic model
 */
export interface ActivityDetail {
  name: string
  description?: string
  category: string
  latitude?: number
  longitude?: number
  start_time?: string
  duration_minutes?: number
}

/**
 * Trip day detail (simplified for responses)
 * Matches backend: TripDayDetail Pydantic model
 */
export interface TripDayDetail {
  day_number: number
  title: string
  trip_date?: string
  activities: ActivityDetail[]
}

/**
 * Trip detail response
 * Matches backend: TripDetail Pydantic model
 */
export interface TripDetail extends TripPublic {
  trip_days: TripDayDetail[]
  total_budget_min?: number
  total_budget_max?: number
  is_public: boolean
  view_count: number
}

/**
 * Trip template for suggested itineraries
 * Matches backend: app/models/analytics.py - TripTemplate
 */
export interface TripTemplate {
  id: number
  title: string
  description?: string
  destination: string
  days: number
  budget?: string
  travel_style: string[]
  icon?: string
  template_data: Record<string, unknown>
  is_active: boolean
  sort_order: number
  created_at: string
}
